docs/usage.md

```diff
@@ -0,0 +1,350 @@
+
+Use this to specify the location of your input folder containing BAM files. For example:
+
+```
+--bam_folder "/path/to/folder/where/bam/files/are"            REQUIRED
+--getBai "true"                                               OPTIONAL  (default: "false")
+```
+
+- In case only some specific files inside the BAM folder should be used as input, a file prefix can be defined by:
+  - `--bam_file_prefix`
+
+```
+--bam_file_prefix MYPREFIX
+```
+
+All the BAM files on which the variant calling should be performed should be all stored in the same folder. If you already have the index files (BAI) they should be stored in the same folder and called with the same prefix as the correspoding BAM file ( e.g. file.bam and file.bam.bai ).
+
+**! TIP**
+All the input files can be used in s3 buckets too and the s3://path/to/files/in/bucket can be used instead of a local path.
+
+### Reference Genomes
+
+### Genome
+
+The pipelines can acccept the refernece genome that was used to create the BAM file(s) in one of two ways. Either the reference genome can be specified eg `--hg19` (default) or by supplying the relevant fasta files generated from the `samtools faidx command`.
+
+By default the hg19 version of the reference genome is used. If you want to use it, you do not have to pass anything.
+
+If you do not want to use the deafult version, here is how it works:
+
+Two standard version of the genome ( hg19 and GRCh38.p10 ) are prepared with all their compressed and indexed file in a lifebit s3 bucket.
+They can be used by using one of the flags:
+
+- hg19 (default)
+- `--hg19`
+- chr20 for testing purposes
+  - `--genome hg19chr20`
+- GRCh38
+  - `--h38`
+- GRCh37 primary
+  - `--grch37primary`
+- hs37d5
+  - `--hs37d5`
+
+OR a user can use an own reference genome version, by using the following parameters:
+
+### Fasta
+
+The following parameter are optional:
+
+- Path to fasta reference
+  - `--fasta`
+- Path to fasta index
+  - `--fai`
+- Path to gzipped fasta
+  - `--fastagz`
+- Path to index of gzipped fasta
+  - `--gzfai`
+- Path to bgzip index format (.gzi)
+  - `--gzi`
+
+If the optional parameters are not passed, they will be automatically be produced for you and you will be able to find them in the "preprocessingOUTPUT" folder.
+
+### Exome Data
+
+- For exome bam files
+  - `--exome`
+- Path to bedfile, can be used for exome data
+  - `--bed`
+
+If you are running on exome data you need to prodive the `--exome` flag so that the right verison of the model will be used.
+Moreover, you can provide a bed file.
+
+```
+nextflow run main.nf --exome --hg19 --bam_folder myBamFolder --bed myBedFile
+```
+
+### Advanced Parameters
+
+#### CPUs
+
+The **make_example** process can be internally parallelized and it can be defined how many cpus should be assigned to this process.
+By default all the cpus of the machine are used.
+
+- Number of cores used by machine for makeExamples (default = all)
+  - `--j`
+
+```
+--j 2          OPTIONAL (default: all)
```

Comment:
Hi @ewels,

I am not familiar with `task.cpus`, am I able to just replace `params.j` with it? 
Also, how do I make sure it's specific to `makeExamples` process?

Thanks in advance

Reply:
Nextflow allows you to specify the required cpus, memory and time for any process (see [docs](https://www.nextflow.io/docs/latest/process.html#cpus)). This can be abstracted into an external config file, so that the defaults can be overwritten with different values for different infrastructures. Currently, in `base.config` you're setting these defaults for _all_ processes in your workflow:
https://github.com/nf-core/deepvariant/blob/187eaf15267ca6938fd0cbf42ec94ee8a6d5ac28/conf/base.config#L16-L18

This means that when launched on a cluster or in the cloud, these resources will be requested for every workflow task. You can then customise those defaults for each process in the pipeline (see eg. [here](https://github.com/nf-core/rnaseq/blob/1cd5ab733a35727b01370de2ae8462d65f57fbb9/conf/base.config#L24-L29)).

In the workflow execution, `task` is then a nextflow groovy variable that holds this information, so `task.cpus` contains the number of cpus that has been assigned.

So for the `makeExamples` process, you can add a new default in `base.config`:

```nextflow
process {
    withName: 'makeExamples': {
        cpus: 2
        // this can be configurable also, for example to use the existing params option:
        // cpus: { params.j }
    }
}
```

Then just reference this in the script:
```
process makeExamples {
    input: //...
    output: //...
    script:
    """
    my_command --cpus ${task.cpus} ...
    """
}
```

We do this all over the place in other pipelines. See for example [here](https://github.com/nf-core/rnaseq/blob/1cd5ab733a35727b01370de2ae8462d65f57fbb9/main.nf#L580).

So there are two things here: firstly, the process requirements should be well defined - big enough that jobs can run quickly and without failing, but not excessive so that resources are wasted on clusters and in the cloud. Secondly, once defined, using these as variables to inform scripts about how many resources they have available to them using the `task` variable.

Reply:
@ewels okay thats great, many thanks for your detailed reply.

Also you mentioned in later comment about the line:
```nextflow
int cores = Runtime.getRuntime().availableProcessors()
```

This is used at Lifebit to ensure that all cpus of a machine are used when executing the pipeline on an AWS instance. Would it be alright if this is left as default behaviour but specified in `base.config`?

Reply:
Not quite - the default config in `base.config` should work for most people out of the box. This isn't the case with this snippet. Stuff for running on lifebit should be in a `lifebit` config file and called with `-profile lifebit` when running the pipeline. Then you can have all of the customisation that you need in there without affecting other users of the pipeline.

I guess you should be able to use that snippet quite easily within this config file though:

```nextflow
process {
    cpus: { Runtime.getRuntime().availableProcessors() }
}
```
The squiggly brackets defer the execution of this code until runtime. It will still be on the head node instead of the compute node I think, but if you run everything on a single machine in local mode then that will be fine. This is of course guesswork without actually trying it though :)

NB: AWS batch is way nicer and integrates perfectly with this resource allocation model 😉

Reply:
Ah nice, will definitely use the brackets to defer the runtime and need to look into AWS batch some more 😄

Reply:
@apeltzer wrote a pretty good introductory blog post about it 😉 https://apeltzer.github.io/post/01-aws-nfcore/ There's also the [original nextflow article](https://www.nextflow.io/blog/2017/scaling-with-aws-batch.html) and a follow-up to Alex's article [here](https://antunderwood.gitlab.io/bioinformant-blog/posts/running_nextflow_on_aws_batch/). Because you used the nf-core template, you have mostly built-in support in this pipeline already too..